diskrsync-tcp/main.go

```diff
@@ -0,0 +1,273 @@
+			fmt.Fprintf(os.Stderr, "control-file must be specified with target flag\n")
+			usage()
+			os.Exit(1)
+		}
+		defer func() {
+			logger.Info("Writing control file", "file", *controlFile)
+			if err := createControlFile(*controlFile); err != nil {
+				logger.Error(err, "Unable to create control file")
+			}
+		}()
+		if err := startServer(os.Args[1], *port, &opts, logger); err != nil {
+			logger.Error(err, "Unable to start server to write to file", "target file", os.Args[1])
+			os.Exit(1)
+		}
+	} else {
+		fmt.Fprintf(os.Stderr, "Either source or target must be defined\n")
+		usage()
+		os.Exit(1)
+	}
+	logger.Info("Successfully completed sync")
+}
+
+func createControlFile(fileName string) error {
+	if err := os.MkdirAll(filepath.Dir(fileName), 0755); err != nil {
+		return err
+	}
+	_, err := os.Create(fileName)
+	return err
+}
+
+func connectToTarget(sourceFile, targetAddress string, port int, opts *options, logger logr.Logger) error {
+	f, err := os.Open(sourceFile)
+	if err != nil {
+		return err
+	}
+	logger.Info("Opened filed", "file", sourceFile)
+	defer f.Close()
+	var src io.ReadSeeker
+
+	// Try to open as an spgz file
+	sf, err := spgz.NewFromFile(f, os.O_RDONLY)
+	if err != nil {
+		if err != spgz.ErrInvalidFormat {
+			return err
+		}
+		logger.Info("Not an spgz file")
+		src = f
+	} else {
+		logger.Info("spgz file")
+		src = sf
+	}
+
+	size, err := src.Seek(0, io.SeekEnd)
+	if err != nil {
+		return err
+	}
+
+	_, err = src.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+
+	conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", targetAddress, port))
+	if err != nil {
+		return err
+	}
+	logger.Info("source", "size", size)
+	calcProgress := &progress{
+		progressType: "calc progress",
+		logger:       logger,
+	}
+	syncProgress := &progress{
+		progressType: "sync progress",
+		logger:       logger,
+	}
+	err = diskrsync.Source(src, size, conn, conn, true, opts.verbose, calcProgress, syncProgress)
+	cerr := conn.Close()
+	if err == nil {
+		err = cerr
+	}
+	return err
+}
+
+func startServer(targetFile string, port int, opts *options, logger logr.Logger) error {
+	var w spgz.SparseFile
+	useReadBuffer := false
+
+	f, err := os.OpenFile(targetFile, os.O_RDWR|os.O_CREATE, 0666)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	logger.Info("Opened file", "file", targetFile)
+	info, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	logger.Info("file info", "info", info)
+
+	if info.Mode()&(os.ModeDevice|os.ModeCharDevice) != 0 {
+		logger.Info("device file?")
+		w = spgz.NewSparseFileWithoutHolePunching(f)
+		useReadBuffer = true
+	} else if !opts.noCompress {
+		sf, err := spgz.NewFromFileSize(f, os.O_RDWR|os.O_CREATE, diskrsync.DefTargetBlockSize)
+		if err != nil {
+			if err != spgz.ErrInvalidFormat {
+				if err == spgz.ErrPunchHoleNotSupported {
+					err = fmt.Errorf("target does not support compression. Try with -no-compress option (error was '%v')", err)
+				}
+				return err
+			}
```

Comment:
what happens here if the error is ErrInvalidFormat? Is this ignored on purpose?

Reply:
Yes it is, in that case it will assume it is just a raw file file and treat it as such.